### ISSUE TITLE
very high latency compared to winsock programming for short msg

### ISSUE THREAD

**Issue by gitpupillo** [zeromq/libzmq#2038]
Hi,
I run local_lat/remote_lat  test first on 127.0.0.1 and then between two machines connected via direct cable. I then made an equivalente test throughbasic  windows winsock programming. Everything with 256bytes messages  and 10000 times roundtrip. Here the results:

ZMQ           127.0.0.1 lat=130uS        two machines lat=620uS
WINSOCK  127.0.0.1  lat=    8uS        two machines lat= 95uS

machine used for 127.0.0.1 test and as client for the 2nd test is:
    i7 2.9Ghz 10gb eth Win 7 64 pro 4GB RAM laptop
machine used only as server in the 2nd test is:
    AMD P320 2.1Ghz  1Gb eth Win 7 Home 64 2GB RAM laptop

I expected a diffrence of course but not such a difference so I don't figure out if it is an issue or not
Thanks Pupillo

**Comment by somdoron** [comment]
Can you try bigger number of repeats?

**Comment by gitpupillo** [comment]
Yes, tried 100000, with the same results. 
But I then built release configurations both for zmq and winsock (they both were debug before).
Whilst for winsock results are the same, for zmq results changed a lot:
always 10000 msg 256B:
ZMQ         127.0.0.1 lat=45uS      two machines lat=203uS
WINSOCK 127.0.0.1 lat=  8uS       two machines lat=  95uS

Much better but still pretty worse than winsock.  Is it suspicious or is it what one could expect?

**Comment by somdoron** [comment]
Zeromq is not tuned for latency but for throughput,  so you should have
latency that is worse than pure tcp.  However on a big batch it should get
closer. The latency test program doesn't actually test big batch...

We can design a test that test with a big batch. Let me know if it would
help.

**Comment by gitpupillo** [comment]
Does batch mean a group of clients in parallel? I built an echo server by ZMQ_ROUTER (no proxy, banal echo)  and a client with 4 threads, each one  connecting to the server via a ZMQ_DEALER socket. Always with 256B, on 127.0.0.1 I got a minimum of 19uS latency, with 8 client threads (2 processes) active. If I run more clients then latency start to increase again. I thought it was because more clients drained CPU from the server.. So I took another machine.So 3 machines connected via cable with a router in the middle.  Now, in one machine only the server running. I start 12 threads (3 processes) in the 2nd machine and I get 47uS latency. Now if I start some processes form the 3rd machine, then latency starts again increasing (55us and more). WHY? Latency is as seen from the server, not from a client, so it should always decrease down to the min achievable when increasing the number of client from another machine. Latency=elapsed/(num_of_pack_echoed*2).
I don't think that it depends on the router as if I directly connect the first two machines I get the same results of the situation with the router in the middle. 

I wonder what should I expect from my final app which expects 20-50000 clients...
